File: 2022/Day2.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.Scanner;

public class Day2 {

    public static int[][] outcomes = {{3, 6, 0},
                                      {0, 3, 6},
                                      {6, 0, 3}};
    public static void main(String[] args) throws FileNotFoundException{
        Day2 solution = new Day2();
        System.out.println("Part 1 Answer: " + solution.part1(new Scanner(new File("input.txt"))));
        System.out.println("Part 2 Answer: " + solution.part2(new Scanner(new File("input.txt"))));
    }

    public int part1(Scanner input) {
        int total = 0;
        while (input.hasNextLine()) {
            String data = input.nextLine();
            int opponent = charToInt(data.charAt(0));
            int player = charToInt(data.charAt(2));
            total += player + 1 + outcomes[opponent][player];
        }
        return total;
    }

    public int part2(Scanner input) {
        int total = 0;
        while (input.hasNextLine()) {
            String data = input.nextLine();
            int opponent = charToInt(data.charAt(0));
            int outcome = charToInt(data.charAt(2)) * 3;
            total += outcome;
            for (int i = 0; i < 3; i++) {
                if (outcomes[opponent][i] == outcome) {
                    total += i + 1;
                    break;
                }
            }
        }
        return total;
    }

    public int charToInt(char c) {
        switch (c) {
            case 'A':
            case 'X':
                return 0;
            case 'B':
            case 'Y':
                return 1;
            default:
                return 2;
        }
    }
}

File: 2023/Day2.cs
class Day2
{
    static void Main()
    {
        string[] data = File.ReadLines("input.txt").ToArray();
        Console.WriteLine("Part 1 Answer: " + Part1(data));
        Console.WriteLine("Part 2 Answer: " + Part2(data));
    }

    static int Part1(string[] data)
    {
        int total = 0;
        for (int i = 0; i < data.Length; i++)
        {
            bool possible = true;
            string[] line = data[i].Substring(7 + (i + 1).ToString().Length).Split("; ");
            foreach (string list in line)
            {
                string[] colourList = list.Split(", ");
                foreach (string countAndColour in colourList)
                {
                    string[] countAndColourList = countAndColour.Split(" ");
                    int count = int.Parse(countAndColourList[0]);
                    string colour = countAndColourList[1];
                    if (count > 12 && colour == "red" || count > 13 && colour == "green" || count > 14 && colour == "blue")
                    {
                        possible = false;
                        break;
                    }
                }
            }
            if (possible)
            {
                total += i + 1;
            }
        }
        return total;
    }

    static int Part2(string[] data)
    {
        int total = 0;
        for (int i = 0; i < data.Length; i++)
        {
            int mostReds = 0;
            int mostGreens = 0;
            int mostBlues = 0;
            string[] line = data[i].Substring(7 + (i + 1).ToString().Length).Split("; ");
            foreach (string list in line)
            {
                string[] colourList = list.Split(", ");
                foreach (string countAndColour in colourList)
                {
                    string[] countAndColourList = countAndColour.Split(" ");
                    int count = int.Parse(countAndColourList[0]);
                    string colour = countAndColourList[1];
                    switch (colour)
                    {
                        case "red":
                            mostReds = Math.Max(mostReds, count);
                            break;
                        case "green":
                            mostGreens = Math.Max(mostGreens, count);
                            break;
                        case "blue":
                            mostBlues = Math.Max(mostBlues, count);
                            break;
                    }
                }
            }
            total += mostReds * mostGreens * mostBlues;
        }
        return total;
    }
}
